File: hermit-cli/tests/mmap_stress_determinism.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! End-to-end determinism coverage for an mmap-heavy guest.
//!
//! `tests/c/mmap_stress_determinism.c` allocates many anonymous maps of varied
//! sizes, touches their pages deterministically, toggles `mprotect`
//! protections, grows and shrinks a region with `mremap`, and drives a
//! `MAP_SHARED` anonymous region through `msync`, then folds every touched byte
//! into an FNV-1a checksum. Because Hermit fixes the guest address layout, both
//! the addresses and the resulting checksum must be identical across runs; the
//! guest prints the checksum and the `mmap_stress success` marker so a stable
//! stdout is the determinism witness.

use std::fs;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::process::Output;
use std::sync::Mutex;
use std::sync::MutexGuard;
use std::sync::OnceLock;

const RUNS: usize = 5;
const SUCCESS_MARKER: &str = "mmap_stress success\n";

static HERMIT_RUN_LOCK: Mutex<()> = Mutex::new(());
static MMAP_STRESS_GUEST: OnceLock<PathBuf> = OnceLock::new();

fn hermit_run_lock() -> MutexGuard<'static, ()> {
    HERMIT_RUN_LOCK
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner())
}

fn command_output(mut command: Command, label: &str) -> Output {
    let rendered = format!("{command:?}");
    let output = command
        .output()
        .unwrap_or_else(|error| panic!("failed to start {label}: {rendered}: {error}"));
    assert!(
        output.status.success(),
        "{label} failed: {rendered}\nstatus: {}\nstdout:\n{}\nstderr:\n{}",
        output.status,
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr),
    );
    output
}

fn mmap_stress_guest() -> &'static Path {
    MMAP_STRESS_GUEST.get_or_init(|| {
        let repository = Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .expect("hermit-cli should be inside the repository");
        let build_root = Path::new(env!("CARGO_TARGET_TMPDIR")).join("mmap-stress-determinism");
        fs::create_dir_all(&build_root)
            .expect("failed to create mmap-stress guest build directory");
        let output = build_root.join("mmap_stress_determinism");

        let mut command = Command::new("cc");
        command
            .args([
                "-O0",
                "-g",
                "-D_GNU_SOURCE",
                "-std=c11",
                "-Wall",
                "-Wextra",
                "-Werror",
            ])
            .arg(repository.join("tests/c/mmap_stress_determinism.c"))
            .arg("-o")
            .arg(&output);
        command_output(command, "mmap-stress guest compilation");
        output
    })
}

fn run_once(run: usize) -> Vec<u8> {
    let mut command = Command::new(env!("CARGO_BIN_EXE_hermit"));
    command
        .args([
            "run",
            "--base-env=minimal",
            "--no-virtualize-cpuid",
            "--max-timeslice=disabled",
            "--",
        ])
        .arg(mmap_stress_guest());

    let output = command_output(command, &format!("mmap-stress run {run}/{RUNS}"));
    assert!(
        output.stdout.ends_with(SUCCESS_MARKER.as_bytes()),
        "mmap-stress omitted its success marker on run {run}/{RUNS}:\n{}",
        String::from_utf8_lossy(&output.stdout),
    );
    output.stdout
}

#[test]
fn mmap_stress_layout_and_checksum_are_deterministic() {
    let _guard = hermit_run_lock();
    let expected = run_once(1);

    for run in 2..=RUNS {
        let actual = run_once(run);
        assert_eq!(
            actual,
            expected,
            "mmap-stress output changed on run {run}/{RUNS}:\nexpected:\n{}actual:\n{}",
            String::from_utf8_lossy(&expected),
            String::from_utf8_lossy(&actual),
        );
    }
}

#[test]
#[ignore = "e2e: requires hermit + mount namespaces"]
fn mmap_stress_reaches_strict_verify_l2() {
    let _guard = hermit_run_lock();
    let mut command = Command::new("timeout");
    command
        .args(["--kill-after", "10s", "120s"])
        .arg(env!("CARGO_BIN_EXE_hermit"))
        .args([
            "--log=info",
            "run",
            "--strict",
            "--verify",
            "--no-virtualize-cpuid",
            "--preemption-timeout=disabled",
            "--",
        ])
        .arg(mmap_stress_guest());

    let output = command_output(command, "mmap-stress strict verification");
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        stdout.contains("Determinism verified") || stderr.contains("Determinism verified"),
        "mmap-stress exited 0 without Hermit's determinism marker\nstdout:\n{stdout}\nstderr:\n{stderr}"
    );
}

File: tests/c/mmap_stress_determinism.c
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

// mmap-heavy determinism guest. Exercises many anonymous maps of varied
// sizes with deterministic page-touch, mprotect toggling, mremap grow/shrink,
// a MAP_SHARED anonymous region with msync, and a final FNV-1a checksum of all
// touched bytes. Under Hermit the address layout and byte content must be
// identical across runs, so the printed checksum is a determinism witness.
// Prints "mmap_stress success" as the final line on success.
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>

static uint64_t fnv1a(const uint8_t *p, size_t n, uint64_t h) {
    for (size_t i = 0; i < n; i++) {
        h ^= p[i];
        h *= 1099511628211ULL;
    }
    return h;
}

int main(void) {
    long pg = sysconf(_SC_PAGESIZE);
    uint64_t h = 1469598103934665603ULL;
    size_t total_pages = 0;

    // A: 64 anonymous maps of growing size; touch every page deterministically.
    for (int i = 1; i <= 64; i++) {
        size_t len = (size_t)i * (size_t)pg;
        uint8_t *m = mmap(NULL, len, PROT_READ | PROT_WRITE,
                          MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
        if (m == MAP_FAILED) {
            perror("mmap");
            return 1;
        }
        for (size_t off = 0; off < len; off += pg) {
            m[off] = (uint8_t)((i * 31 + off) & 0xff);
            m[off + pg - 1] = (uint8_t)((i * 17 + off) & 0xff);
            total_pages++;
        }
        if (mprotect(m, len, PROT_READ) != 0) {
            perror("mprotect-ro");
            return 1;
        }
        h = fnv1a(m, len, h);
        if (mprotect(m, len, PROT_READ | PROT_WRITE) != 0) {
            perror("mprotect-rw");
            return 1;
        }
        munmap(m, len);
    }

    // B: mremap grow then shrink, preserving content.
    size_t base = 4 * (size_t)pg;
    uint8_t *r = mmap(NULL, base, PROT_READ | PROT_WRITE,
                      MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (r == MAP_FAILED) {
        perror("mmap-r");
        return 1;
    }
    memset(r, 0xab, base);
    uint8_t *r2 = mremap(r, base, base * 4, MREMAP_MAYMOVE);
    if (r2 == MAP_FAILED) {
        perror("mremap-grow");
        return 1;
    }
    memset(r2 + base, 0xcd, base * 3);
    h = fnv1a(r2, base * 4, h);
    uint8_t *r3 = mremap(r2, base * 4, base, MREMAP_MAYMOVE);
    if (r3 == MAP_FAILED) {
        perror("mremap-shrink");
        return 1;
    }
    h = fnv1a(r3, base, h);
    munmap(r3, base);

    // C: MAP_SHARED anonymous region + msync.
    size_t slen = 16 * (size_t)pg;
    uint8_t *s = mmap(NULL, slen, PROT_READ | PROT_WRITE,
                      MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if (s == MAP_FAILED) {
        perror("mmap-shared");
        return 1;
    }
    for (size_t off = 0; off < slen; off++) {
        s[off] = (uint8_t)(off * 7 + 3);
    }
    if (msync(s, slen, MS_SYNC) != 0) {
        perror("msync");
        return 1;
    }
    h = fnv1a(s, slen, h);
    munmap(s, slen);

    printf("total_pages=%zu\n", total_pages);
    printf("checksum=%016llx\n", (unsigned long long)h);
    printf("mmap_stress success\n");
    return 0;
}
